File: src/solution_java/medium/_89GrayCode.java
package solution_java.medium;

import java.util.ArrayList;
import java.util.List;

public class _89GrayCode {
    public class Solution {
        public List<Integer> grayCode(int n) {
            List<Integer> result = new ArrayList<>();
            result.add(0);
            for (int i = 1; i <= n; i++) {
                int length = result.size();
                for (int j = length - 1; j >= 0; j--) {
                    result.add(result.get(j) | (1 << (i - 1)));
                }
            }
            return result;
        }
    }
}

File: src/medium/_89GrayCode.kt
package medium

class _89GrayCode {
    class Solution {
        fun grayCode(n: Int): List<Int> {
            val result = ArrayList<Int>()
            result.add(0)
            for (i in 0 until n) {
                val high = 1 shl i
                val length = result.size - 1
                for (j in length downTo 0) {
                    val newItem = high + result[j]
                    result.add(newItem)
                }
            }
            return result
        }
    }
}